[Swiftify] Add support for free functions imported as instance methods

hnrklssn opened the pull request:
This adds support for attaching safe interop wrappers to functions explicitly imported as instance methods using swift_name.

rdar://156288883

hnrklssn committed:
[Swiftify] Add support for free functions imported as instance methods

This adds support for attaching safe interop wrappers to functions
explicitly imported as instance methods using swift_name.

rdar://156288883

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented:
@swift-ci please smoke test

Xazax-hun reviewed (approved):
Looks great, left some small questions/nits inline.

DougGregor reviewed (approved):
Thank you!

hnrklssn commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -9519,14 +9532,42 @@ void ClangImporter::Implementation::swiftify(AbstractFunctionDecl *MappedDecl) {
+        SwiftifyInfoPrinter::SELF_PARAM_INDEX;
+      ParamDecl *swiftParam = nullptr;
+      if (mappedIndex == SwiftifyInfoPrinter::SELF_PARAM_INDEX) {
+        swiftParam = MappedDecl->getImplicitSelfDecl(/*createIfNeeded*/true);
```
Yup! Not sure what controls when it's created though.

hnrklssn committed:
fix tests in CI

hnrklssn committed:
add namespace tests

hnrklssn committed:
use swift/bridging in test

hnrklssn committed:
[Swiftify] properly forward inout parameters using `&`

While we handled prepending & to MutableSpan parameters, regular
parameters that were unchanged during the transformation were not
checked for inout-ness.

hnrklssn commented on test/Interop/Cxx/swiftify-import/import-as-instance-method.swift:
```diff
@@ -57,6 +59,13 @@ const IntSpan spanConstSelf(const IntSpan p __lifetimebound) __attribute__((swif
 ConstIntSpan constSpanSelf(ConstIntSpan p __lifetimebound) __attribute__((swift_name("ConstIntSpan.constSpanSelf(self:)")));
 
+namespace baz {
+  void namespaced(A *a, IntSpan p __noescape) __attribute__((swift_name("A.namespaced(self:_:)")));
```
Done. Found a bug with forwarding of `inout` parameters in the process!

hnrklssn commented:
@swift-ci please smoke test

hnrklssn committed:
try to debug CI diff failure

hnrklssn committed:
[Swiftify] Reorder RUN lines (NFC)

Since the compiler invokation performing the macro expansion dump
redirects its stderr output, any errors are not displayed by default.
This is extra problematic for test failures in CI. By performing the
compiler invokation with -verify first, errors are shown.

hnrklssn commented:
@swift-ci please smoke test

hnrklssn committed:
deduplicate path

hnrklssn commented:
@swift-ci please smoke test

hnrklssn committed:
re-export std::span regardless of module name

hnrklssn commented:
@swift-ci please smoke test

Xazax-hun commented on lib/Macros/Sources/SwiftMacros/SwiftifyImportMacro.swift:
```diff
@@ -468,7 +468,14 @@ struct FunctionCallBuilder: BoundsCheckedThunkBuilder {
     let args: [ExprSyntax] = base.signature.parameterClause.parameters.enumerated()
       .map { (i: Int, param: FunctionParameterSyntax) in
-        return pointerArgs[i] ?? ExprSyntax("\(param.name)")
+        if let overrideArg = pointerArgs[i] {
```
I wonder if this is a fix we want to backport to 6.2.

hnrklssn replied:
Yeah that's probably a good idea

hnrklssn commented:
@swift-ci please smoke test

hnrklssn committed:
add std::span to module map for libstdc++

hnrklssn commented:
@swift-ci please smoke test

hnrklssn committed:
fix -verify-additional-file path on windows

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented:
@swift-ci please smoke test linux
